Code/GraphMol/Descriptors/AtomicEnvironmentVector.cpp

```diff
@@ -0,0 +1,679 @@
+      }
+      index += (m * (m - 1) / 2);
+    }
+    index = 0;
+
+    for (auto i = intraPairIndices.rows() / 2; i < intraPairIndices.rows();
+         i++) {
+      for (int j = 0; j < m * (m - 1) / 2; j++) {
+        intraPairIndicesFlattened(1, index + j) = intraPairIndices(i, j);
+      }
+      index += (m * (m - 1) / 2);
+    }
+  }
+  ArrayXXd sortedLocalIndex12(2, (mask.array() == 1).count());
+  index = 0;
+  for (auto i = 0; i < mask.size(); i++) {
+    if (mask(0, i) == 1) {
+      sortedLocalIndex12.col(index) << intraPairIndicesFlattened.col(i);
+      index++;
+    }
+  }
+
+  auto cumsumCount = cumsumFromZero(counts);
+
+  VectorXd extraLocalIndex12(pairIndices.size());
+  index = 0;
+  for (auto i : pairIndices) {
+    extraLocalIndex12(index) = cumsumCount[i];
+    index++;
+  }
+
+  sortedLocalIndex12 = (sortedLocalIndex12.matrix().rowwise() +
+                          extraLocalIndex12.transpose())
+                             .array();
+
+  ArrayXXd localIndex12(2, sortedLocalIndex12.cols());
+  for (int j = 0; j < sortedLocalIndex12.cols(); j++) {
+    localIndex12(0, j) = revIndices[sortedLocalIndex12(0, j)];
+    localIndex12(1, j) = revIndices[sortedLocalIndex12(1, j)];
+  }
+
+  return std::make_pair(centralAtomIndex, localIndex12);
+}
+
+ArrayXXd TriuIndex(int numSpecies) {
+  std::vector<int> species1, species2, pairIndex;
+
+  for (auto i = 0; i < numSpecies; i++) {
+    for (int j = i; j < numSpecies; j++) {
+      species1.push_back(i);
+      species2.push_back(j);
+    }
+    pairIndex.push_back(i);
+  }
+  ArrayXXd ret(numSpecies, numSpecies);
+  int index1 = 0;
+
+  for (size_t i = 0; i < species1.size(); i++) {
+    ret(species1[i], species2[i]) = index1;
+    ret(species2[i], species1[i]) = index1;
+    index1++;
+  }
+
+  return ret;
+}
+
+//! Accumulate the elements of a Matrix into another Matrix by adding to the
+//! indices in the order given in index.
+/*!
+  \param vector1    Matrix to which values are to be added
+  \param vector2    Matrix from which values are to be added
+  \param index      Indices in order of which values are added
+  \param multi      Number of pairs to be considered
+  \param numAtoms   Number of atoms in the molecules
+
+  \return Matrix containing accumulated elements of vector2 into vector1
+  according to order given in index
+
+  \verbatim
+    Index[i] == j, then the ith row of vector2 is added to the jth row of
+  vector1
+  \endverbatim
+*/
+ArrayXXd IndexAdd(ArrayXXd vector1, ArrayXXd vector2, ArrayXXd index, int multi,
+                  int numAtoms) {
+  for (int idxCol = 0; idxCol < index.cols(); idxCol++) {
+    for (auto i = 0; i < index.rows(); i++) {
+      for (int j = 0; j < multi * numAtoms; j++) {
+        if (index(i, idxCol) == j) {
+          vector1.row(j) += vector2.row(i);
+        }
+      }
+    }
+  }
+  return vector1;
+}
+
+ArrayXXd AtomicEnvironmentVector(const ROMol &mol, int confId) {
```

Comment:
Again, this is a function that contains a lot of code, but very little in the way of inline comments. Having a bit of text in there to explain what's going on would help with review and maintenance.